public/app/plugins/datasource/loki/formatterTS.ts

```diff
@@ -0,0 +1,406 @@
+        lastPipelineType = JsonExpressionParser;
+        break;
+
+      case LabelFilter:
+        response += buildResponse(LabelFilter, lastPipelineType, formatLabelFilter(node, query));
+        lastPipelineType = LabelFilter;
+        break;
+
+      case LineFormatExpr:
+        response += buildResponse(LineFormatExpr, lastPipelineType, formatLineFormatExpr(node, query));
+        lastPipelineType = LineFormatExpr;
+        break;
+
+      case LabelFormatExpr:
+        response += buildResponse(LabelFormatExpr, lastPipelineType, formatLabelFormatExpr(node, query));
+        lastPipelineType = LabelFormatExpr;
+        break;
+
+      case LineComment:
+        response += buildResponse(LineComment, lastPipelineType, formatLineComment(node, query));
+        lastPipelineType = LineComment;
+        break;
+    }
+  });
+
+  return response;
+}
+
+export function formatLineFilter(node: SyntaxNode, query: string): string {
+  const filterNode = node.getChild(Filter)!;
+  const filterOperation = node.getChild(FilterOp);
+  const stringNode = node.getChild(String)!;
+
+  const filter = query.substring(filterNode.from, filterNode.to);
+  const string = query.substring(stringNode.from, stringNode.to);
+
+  if (filterOperation) {
+    return `${filter} ip(${string})`;
+  }
+  return `${filter} ${string}`;
+}
+
+export function formatLabelParser(node: SyntaxNode, query: string): string {
+  const hasString = node.getChild(String);
+
+  if (hasString) {
+    const parserNode = (node.getChild(Regexp) || node.getChild(Pattern))!;
+    const stringNode = node.getChild(String)!;
+
+    const parser = query.substring(parserNode.from, parserNode.to);
+    const string = query.substring(stringNode.from, stringNode.to);
+
+    return `| ${parser}${string}`;
+  }
+
+  const labelParser = query.substring(node.from, node.to);
+  return `| ${labelParser}`;
+}
+
+export function formatJsonExpressionParser(node: SyntaxNode, query: string): string {
+  const jsonExpressionNodes = iterateNode(node, [JsonExpression]);
+  let response = '';
+
+  jsonExpressionNodes.forEach((node) => {
+    const identifierNode = node.getChild(Identifier)!;
+    const valueNode = node.getChild(String)!;
+
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    response += `${identifier}=${value}, `;
+  });
+
+  return `| json ${trimEnd(response, ', ')}`;
+}
+
+export function formatLabelFilter(node: SyntaxNode, query: string): string {
+  if (node.getChild(Matcher)) {
+    const matcherNode = node.getChild(Matcher)!;
+
+    const identifierNode = matcherNode.getChild(Identifier)!;
+    const valueNode = matcherNode.getChild(String)!;
+    const identifier = query.substring(identifierNode.from, identifierNode.to);
+    const value = query.substring(valueNode.from, valueNode.to);
+
+    if (matcherNode.getChild(Eq)) {
```

Comment:
Here you could also use something like `nextSibling` and avoid the literal check.